server/filestore.go

```diff
@@ -5299,23 +5300,38 @@ func (fs *fileStore) expireMsgs() {
 	var smv StoreMsg
 	var sm *StoreMsg
 	fs.mu.RLock()
+	eflr := fs.eflr
 	maxAge := int64(fs.cfg.MaxAge)
 	minAge := time.Now().UnixNano() - maxAge
 	fs.mu.RUnlock()
 
 	if maxAge > 0 {
-		for sm, _ = fs.msgForSeq(0, &smv); sm != nil && sm.ts <= minAge; sm, _ = fs.msgForSeq(0, &smv) {
+		for sm, _ = fs.msgForSeq(eflr, &smv, true); sm != nil && sm.ts <= minAge; sm, _ = fs.msgForSeq(eflr, &smv, true) {
+			if len(sm.hdr) > 0 {
+				if ttl, err := getMessageTTL(sm.hdr); err == nil && ttl < 0 {
+					// The message has a negative TTL, therefore it must "never expire".
+					minAge = time.Now().UnixNano() - maxAge
+					eflr = sm.seq + 1
+					continue
+				}
+			}
 			fs.mu.Lock()
 			fs.removeMsgViaLimits(sm.seq)
 			fs.mu.Unlock()
 			// Recalculate in case we are expiring a bunch.
 			minAge = time.Now().UnixNano() - maxAge
+			eflr = sm.seq + 1
```

Comment:
Note that when we do this, `sm.seq + 1` may not actually exist and we could stumble across a gap, as it could have been deleted. Hence the need to add the `next` flag into `msgForSeq` to scan for the next sequence that _does_ exist. 

Without this, it's not possible to have messages A, B and C where A has `Nats-TTL: never`, B has maybe been deleted and C has any other TTL.